src/network/core/packet.cpp

```diff
@@ -337,3 +342,33 @@ void Packet::Recv_string(char *buffer, size_t size, StringValidationSettings set
 
 	str_validate(bufp, last, settings);
 }
+
+/**
+ * Get the amount of bytes that are still available for the Transfer functions.
+ * @return The number of bytes that still have to be transfered.
+ */
+size_t Packet::RemainingBytesToTransfer() const
+{
+	return this->size - this->pos;
+}
+
+/**
+ * Transfer data from the given function into the packet starting writing at the
+ * position the last transfer stopped.
+ * This will attempt to write all the remaining bytes into the packet, and updates
+ * the position accordingly to how many bytes were actually written by the called
+ * transfer_function.
+ * @param transfer_function The function to pass the char *buffer and size to read,
+ *                          which returns the amount that was read or -1 upon errors.
+ * @return The return value of the transfer_function.
+ */
+ssize_t Packet::TransferIn(std::function<ssize_t(char*, size_t)> transfer_function)
+{
+	size_t amount = this->RemainingBytesToTransfer();
+	if (amount == 0) return 0;
+
+	char *input_buffer = reinterpret_cast<char*>(&this->buffer[this->pos]);
```

Comment:
In the last cpmmit you change `buffer` to a `std::vector`. Again, this is my lack of C++, but is this still safe in that case? Is a `std::vector` valid to be used like this?

Reply:
One of the guarantees of std::vector is that all elements are laid out in order in a contiguous memory block, such that you can have a pointer to one element and increment the pointer to reach the following element. So yes std::vector does allow this.

Since C++11 there is also a std::vector::data() function that gives a straight pointer to the first element of the underlying array, without having to do gymnastics.
https://en.cppreference.com/w/cpp/container/vector/data

Reply:
I changed some types around so I did not need reinterpret_cast anymore and it uses ::data() now.